builtin/providers/cloudstack/resource_cloudstack_ipaddress.go

```diff
@@ -138,10 +153,21 @@ func resourceCloudStackIPAddressDelete(d *schema.ResourceData, meta interface{})
 func verifyIPAddressParams(d *schema.ResourceData) error {
 	_, network := d.GetOk("network_id")
 	_, vpc := d.GetOk("vpc_id")
+	_, zone := d.GetOk("zone_id")
+
+	f := func(bs ...bool) int {
+		cnt := 0
+		for _, b := range bs {
+			if b {
+				cnt++
+			}
+		}
+		return cnt
+	}
 
-	if (network && vpc) || (!network && !vpc) {
+	if f(network, vpc, zone) >= 2 || (!network && !vpc && !zone) {
 		return fmt.Errorf(
-			"You must supply a value for either (so not both) the 'network_id' or 'vpc_id' parameter")
+			"You must supply one value for either (so not more than two) the 'network_id' or 'vpc_id' or 'zone_id'  parameter")
```

Comment:
I would suggest to keep this function as it was. The `zone_id` is optional so still only one of `network` or `vpc` is needed. The fact that `zone_id` is required by your cloud provider seems to be a somewhat wierd choice they made, which makes them divert from standard CloudStack. So I'm good with adding this feature, but we should not make it required.